libkbfs/disk_block_cache.go

```diff
@@ -0,0 +1,316 @@
+	}
+	return cache.config.Codec().Encode(&entry)
+}
+
+// Get implements the DiskBlockCache interface for DiskBlockCacheStandard.
+func (cache *DiskBlockCacheStandard) Get(ctx context.Context, tlfID tlf.ID,
+	blockID kbfsblock.ID) (buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf, err error) {
+	cache.lock.RLock()
+	defer cache.lock.RUnlock()
+	if cache.isClosed {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{},
+			DiskCacheClosedError{"Get"}
+	}
+	defer func() {
+		cache.log.CDebugf(ctx, "Cache Get id=%s tlf=%s bSize=%d err=%+v", blockID, tlfID, len(buf), err)
+	}()
+	var entry []byte
+	err = runUnlessCanceled(ctx, func() error {
+		blockKey := blockID.Bytes()
+		buf, err := cache.blockDb.Get(blockKey, nil)
+		if err != nil {
+			return NoSuchBlockError{blockID}
+		}
+		err = cache.updateLruLocked(tlfID, blockKey)
+		if err != nil {
+			return err
+		}
+		entry = buf
+		return nil
+	})
+	if err != nil {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, err
+	}
+	return cache.decodeBlockCacheEntry(entry)
+}
+
+// Put implements the DiskBlockCache interface for DiskBlockCacheStandard.
+func (cache *DiskBlockCacheStandard) Put(ctx context.Context, tlfID tlf.ID,
+	blockID kbfsblock.ID, buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf) error {
+	cache.lock.RLock()
+	defer cache.lock.RUnlock()
+	if cache.isClosed {
+		return DiskCacheClosedError{"Put"}
+	}
+	blockLen := len(buf)
+	// TODO: accounting
+	entry, err := cache.encodeBlockCacheEntry(buf, serverHalf)
+	encodeLen := len(entry)
+	cache.log.CDebugf(ctx, "Cache Put id=%s tlf=%s bSize=%d entrySize=%d err=%+v", blockID, tlfID, blockLen, encodeLen, err)
+	if err != nil {
+		return err
+	}
+	blockKey := blockID.Bytes()
+	err = cache.blockDb.Put(blockKey, entry, nil)
+	if err != nil {
+		return err
+	}
+	return cache.updateLruLocked(tlfID, blockKey)
+}
+
+// Delete implements the DiskBlockCache interface for DiskBlockCacheStandard.
+func (cache *DiskBlockCacheStandard) Delete(ctx context.Context, tlfID tlf.ID,
+	blockIDs []kbfsblock.ID) error {
+	cache.lock.RLock()
+	defer cache.lock.RUnlock()
+	if cache.isClosed {
+		return DiskCacheClosedError{"Delete"}
+	}
+	if len(blockIDs) == 0 {
+		return nil
+	}
+	cache.log.CDebugf(ctx, "Cache Delete tlf=%s numBlocks=%d", tlfID, len(blockIDs))
+	blockBatch := new(leveldb.Batch)
+	lruBatch := new(leveldb.Batch)
+	for _, id := range blockIDs {
+		blockKey := id.Bytes()
+		blockBatch.Delete(blockKey)
+		lruKey := cache.lruKey(tlfID, blockKey)
+		lruBatch.Delete(lruKey)
+	}
+	if err := cache.blockDb.Write(blockBatch, nil); err != nil {
+		return err
+	}
+	if err := cache.lruDb.Write(lruBatch, nil); err != nil {
+		return err
+	}
+
+	cache.compactCachesLocked(ctx)
+
+	return nil
+}
+
+// evictLocked evicts a number of blocks from the cache.
+func (cache *DiskBlockCacheStandard) evictLocked(ctx context.Context,
+	tlfID tlf.ID, numBlocks int) error {
+	// Use kbfscrypto.MakeTemporaryID() to create a random hash ID. Then begin
```

Comment:
FWIW I don't want to change this comment at this point because I already have code dependent on it that I'd rather not conflict.